paparazzi/api/paparazzi.api

```diff
@@ -0,0 +1,255 @@
+	public fun equals (Ljava/lang/Object;)Z
+	public final fun getDensity ()Lcom/android/resources/Density;
+	public final fun getFolderConfiguration ()Lcom/android/ide/common/resources/configuration/FolderConfiguration;
+	public final fun getFontScale ()F
+	public final fun getHardwareConfig ()Lcom/android/ide/common/rendering/api/HardwareConfig;
+	public final fun getKeyboard ()Lcom/android/resources/Keyboard;
+	public final fun getKeyboardState ()Lcom/android/resources/KeyboardState;
+	public final fun getLayoutDirection ()Lcom/android/resources/LayoutDirection;
+	public final fun getLocale ()Ljava/lang/String;
+	public final fun getNavigation ()Lcom/android/resources/Navigation;
+	public final fun getNightMode ()Lcom/android/resources/NightMode;
+	public final fun getOrientation ()Lcom/android/resources/ScreenOrientation;
+	public final fun getRatio ()Lcom/android/resources/ScreenRatio;
+	public final fun getReleased ()Ljava/lang/String;
+	public final fun getScreenHeight ()I
+	public final fun getScreenRound ()Lcom/android/resources/ScreenRound;
+	public final fun getScreenWidth ()I
+	public final fun getSize ()Lcom/android/resources/ScreenSize;
+	public final fun getSoftButtons ()Z
+	public final fun getTouchScreen ()Lcom/android/resources/TouchScreen;
+	public final fun getUiMode ()Lcom/android/resources/UiMode;
+	public final fun getUiModeMask ()I
+	public final fun getXdpi ()I
+	public final fun getYdpi ()I
+	public fun hashCode ()I
+	public fun toString ()Ljava/lang/String;
+}
+
+public final class app/cash/paparazzi/DeviceConfig$Companion {
+	public final fun getEnumMap (Ljava/io/File;)Ljava/util/Map;
+	public final fun loadProperties (Ljava/io/File;)Ljava/util/Map;
+}
+
+public final class app/cash/paparazzi/Environment {
+	public static final field $stable I
+	public fun <init> (Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;ILjava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;)V
+	public final fun component1 ()Ljava/lang/String;
+	public final fun component10 ()Ljava/util/List;
+	public final fun component2 ()Ljava/lang/String;
+	public final fun component3 ()Ljava/lang/String;
+	public final fun component4 ()I
+	public final fun component5 ()Ljava/util/List;
+	public final fun component6 ()Ljava/util/List;
+	public final fun component7 ()Ljava/util/List;
+	public final fun component8 ()Ljava/util/List;
+	public final fun component9 ()Ljava/util/List;
+	public final fun copy (Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;ILjava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;)Lapp/cash/paparazzi/Environment;
+	public static synthetic fun copy$default (Lapp/cash/paparazzi/Environment;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;ILjava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;Ljava/util/List;ILjava/lang/Object;)Lapp/cash/paparazzi/Environment;
+	public fun equals (Ljava/lang/Object;)Z
```

Comment:
does equals/hashcode make sense on this?

Reply:
it's part of the data class codegen

Reply:
I know it is, I'm questioning the use of the (data class) tool, not it's result :)

Reply:
Ah got it.  I can't think of any use case; it came for free with the data class codegen, which was added in the first place for the nice `copy` semantics.

Earlier, I spiked out what things would look like by removing `data` and manually writing `toString` and `copy`.  Simple enough, but it's prone to forgetting adding new members in the future.

I'm inclined to leave for now, and maybe move this to a Builder/newBuilder pattern later.  Thoughts?

Reply:
Use Poko

Reply:
Poko helps with the `toString`, but not the `copy`/builder semantics (yet???).

Reply:
Ah yeah. I kinda read the thread backwards so my mind ended up on equals/hashCode.

You have to implement copy yourself since for every new property you must add a new overload.

Reply:
discussed offline.  plan is to go with manual copy + poko + manual equals/hashcode to remove the componentN functions.  a possible additional improvement will be to update poko to allow selectivity of which methods to generate.

Reply:
@drewhamilton @JakeWharton @TWiStErRob PTAAL at the Poko changes

Reply:
LGTM. Similar to when it was a data class, the generated `equals` and `hashCode` aren't really needed, but aren't a compatibility concern and don't hurt anything by being there.

Reply:
You missed `DeviceConfig.released` from `copy`, see https://github.com/cashapp/paparazzi/compare/44a138a58f960c49c322ddd87e0e4f9fbee18dbb..11367688e8275ea34b2b60d6a3b75f6a71c41efe#diff-c2bf71ab4fcca2f6780e9537deb521cc35dba6a7f9a26ee4ed8ff860dce9514fR31 vs 51 on the left.

Reply:
that was intentional, because I didn't think it made sense to copy that member over.  does that intention violate a best practice?

Reply:
I think it violates the principle of least astonishment. `DeviceConfig.PIXEL_XL.copy(locale = X)` would suddenly take the release date of Nexus 4.

Reply:
oh i see what my mistake was: [b2cb074](https://github.com/cashapp/paparazzi/pull/1264/commits/b2cb074ef3d0fcd0b7b9d57f3170df939801f1ef)